exchange/exchange.go

```diff
@@ -305,6 +312,131 @@ func (e *exchange) buildBidResponse(ctx context.Context, liveAdapters []openrtb_
+	if err != nil {
+		return res, err
+	}
+
+	publisher := brandCatExt.Publisher
+
+	seatBidsToRemove := make([]openrtb_ext.BidderName, 0)
+
+	for bidderName, seatBid := range *seatBids {
+		bidsToRemove := make([]int, 0)
+		for bidInd := range seatBid.bids {
+			bid := seatBid.bids[bidInd].bid
+			var tempRespBidExt ResponseBidExt
+			if len(bid.Ext) > 0 {
+				fmt.Println(string(bid.Ext))
+				if err := json.Unmarshal(bid.Ext, &tempRespBidExt); err != nil {
+					return res, err
+				}
+				//if the bid already consists bid.ext.prebid.category then it implies that adapter has returned the final Ad sever category.
+				//if not then look at the bid.cat[0] and translate it to the mapped Ad server category
+				if len(tempRespBidExt.Prebid.Targeting.HbPbCatDur) != 0 {
+					continue
+				}
+			}
+
+			bidIabCat := bid.Cat
+			if len(bidIabCat) != 1 {
+				//on receiving bids from adapters if no unique IAB category is returned  or if no ad server category is returned discard the bid
+				bidsToRemove = append(bidsToRemove, bidInd)
+			} else {
+				//if unique IAB category is present then translate it to the adserver category based on mapping file
+
+				//if mapping required but no mapping file is found then discard the bid
+				fmt.Println(primaryAdServer, publisher)
+
+				cat, err := categoriesFetcher.FetchCategories(primaryAdServer, publisher, bidIabCat[0])
+				if err != nil {
+					//if mapping required but no mapping file is found then discard the bid
+					bidsToRemove = append(bidsToRemove, bidInd)
+				}
+
+				finalCategory := fmt.Sprintf("%.2f_%s_%ds", bid.Price, cat, 30) //TODO: duration
+				res[bid.ID] = finalCategory
+			}
+
+		}
+
+		if len(bidsToRemove) > 0 {
+			if len(bidsToRemove) == len(seatBid.bids) {
+				//if all bids are invalid - remove entire seat bid
+				seatBidsToRemove = append(seatBidsToRemove, bidderName)
+			} else {
+				bids := seatBid.bids
+				for _, bidInd := range bidsToRemove {
+					bids = append(bids[:bidInd], bids[bidInd+1:]...)
+				}
+				seatBid.bids = bids
+			}
+		}
+
+	}
+	if len(seatBidsToRemove) > 0 {
+		if len(seatBidsToRemove) == len(*seatBids) {
+			//delete all seat bids
+			*seatBids = nil
+		} else {
+			for _, seatBidInd := range seatBidsToRemove {
+				delete(*seatBids, seatBidInd)
+			}
+
+		}
+	}
+
+	return res, nil
+}
+
+type ResponseBidExt struct {
+	Prebid ResponsePrebidExt `json:"prebid"`
+	//Bidder json.RawMessage   `json:"bidder"`
+}
+
+type ResponsePrebidExt struct {
+	Targeting Targeting `json:"targeting"`
+	//Type      string    `json:"type"`
+}
+
+type Targeting struct {
+	//HbBidder   string `json:"hb_bidder"`
+	//HbPb       string `json:"hb_pb"`
+	//HbSize     string `json:"hb_size"`
+	HbPbCatDur string `json:"hb_pb_cat_dur"`
+
+	//HbDeal string `json:"hb_deal"`
+	//HbCacheId string `json:"hb_cache_id"`
+	//HbUUID string `json:"hb_uuid"`
+	//MobileApp string `json:"mobile-app"`
+}
+
+func getPrimaryAdSetver(adServerId int) (string, error) {
```

Comment:
Typo in function name - Setver instead of Server

Reply:
Done!